test/e2e/balancing.go

```diff
@@ -0,0 +1,200 @@
+	// This is the server that will serve traffic, including the hostname.
+	rc := &api.ReplicationController{
+		ObjectMeta: api.ObjectMeta{
+			Name: name,
+		},
+		Spec: api.ReplicationControllerSpec{
+			Replicas: replicas,
+			Selector: svc.Spec.Selector,
+			Template: &api.PodTemplateSpec{
+				ObjectMeta: api.ObjectMeta{
+					Labels: svc.Spec.Selector,
+				},
+				Spec: api.PodSpec{
+					Containers: []api.Container{
+						{
+							Name:  name,
+							Image: "gcr.io/google_containers/nettest:1.3",
+							Ports: []api.ContainerPort{{ContainerPort: 8080}},
+						},
+					},
+				},
+			},
+		},
+	}
+	By(fmt.Sprintf("Creating server rc %s in namespace %s", rc.Name, ns))
+	_, err := c.ReplicationControllers(ns).Create(rc)
+	expectNoError(err, fmt.Sprintf("creating rc %s", rc.Name))
+	defer func() {
+		// Resize the replication controller to zero to get rid of pods.
+		By("Cleaning up the replication controller")
+		rcReaper, err := kubectl.ReaperFor("ReplicationController", c)
+		if err != nil {
+			Logf("Failed to cleanup replication controller %v: %v.", rc.Name, err)
+		}
+		if _, err = rcReaper.Stop(ns, rc.Name, nil); err != nil {
+			Logf("Failed to stop replication controller %v: %v.", rc.Name, err)
+		}
+	}()
+
+	selector := labels.SelectorFromSet(rc.Spec.Selector)
+	By("Waiting for replicas to be created.")
+	wait.Poll(time.Second*5, time.Second*60, func() (bool, error) {
+		list, err := c.Pods(ns).List(selector, fields.Everything())
+		if err != nil {
+			fmt.Printf("unexpected error listing pods: %v", err)
+			return false, nil
+		}
+		running := 0
+		for _, pod := range list.Items {
+			if pod.Status.Phase == api.PodRunning {
+				running = running + 1
+			}
+		}
+		return running == rc.Spec.Replicas, nil
+	})
+
+	By("Creating service for replicas")
+	defer func() {
+		err := c.Services(ns).Delete(svc.Name)
+		expectNoError(err, "deleting service")
+	}()
+	if _, err := c.Services(ns).Create(svc); err != nil {
+		expectNoError(err, "creating service")
+	}
+
+	// This pod repeatedly curls the replicated server via the service, and logs the output.
+	testPod := &api.Pod{
+		ObjectMeta: api.ObjectMeta{
+			Name: "tester",
+		},
+		Spec: api.PodSpec{
+			Containers: []api.Container{
+				{
+					Name:    "tester",
+					Image:   "busybox",
+					Command: []string{"sh", "-c", "while true; do wget -qO- http://service:8080/read; sleep 1; done"},
+				},
+			},
+		},
+	}
+	By(fmt.Sprintf("Creating tester pod %s in namespace %s", testPod.Name, ns))
+	_, err = c.Pods(ns).Create(testPod)
+	expectNoError(err, fmt.Sprintf("creating pod %s", testPod.Name))
+
+	// At the end of the test, clean up by removing the pod.
+	defer func() {
+		By("Deleting the tester pod")
+		c.Pods(ns).Delete(testPod.Name, nil)
+	}()
+
+	err = waitForPodRunningInNamespace(c, testPod.Name, ns)
+	expectNoError(err, "waiting for tester pod to start")
+
+	// Wait for us to accumulate a bunch of logs.
+	// TODO: actually call log multiple times until we have 10 hostnames?
```

Comment:
You'll at least want to make sure it's been called three times, otherwise this'll be pretty vulnerable to flaking.